src/solvers/flattening/pointer_logic.cpp

```diff
@@ -99,8 +99,33 @@ exprt pointer_logict::pointer_expr(
   const exprt &object_expr=objects[pointer.object];
 
   exprt deep_object=object_rec(pointer.offset, type, object_expr);
+  const address_of_exprt base(deep_object);
 
-  return address_of_exprt(deep_object, type);
+  if(pointer.offset == 0)
+    return typecast_exprt::conditional_cast(base, type);
+
+  const auto object_size = pointer_offset_size(deep_object.type(), ns);
+  if(object_size.has_value() && *object_size <= 1)
+  {
+    return typecast_exprt::conditional_cast(
+      plus_exprt(base, from_integer(pointer.offset, pointer_diff_type())),
+      type);
+  }
+  else if(object_size.has_value() && pointer.offset % *object_size == 0)
+  {
+    return typecast_exprt::conditional_cast(
+      plus_exprt(
+        base, from_integer(pointer.offset % *object_size, pointer_diff_type())),
+      type);
```

Comment:
You probably meant pointer.offset / *object_size

Reply:
Ooops - that clearly shows that my test case is not good enough. Thanks for catching that! I'll extend my test.